Don’t open “details” when interacting with interactive “summary” descendants

sideshowbarker opened the pull request:
#### aed2f1dba4a42f57914f579efc8f76b7c9fe8f3e
<pre>
Don’t open “details” when interacting with interactive “summary” descendants
<a href="https://bugs.webkit.org/show_bug.cgi?id=218758">https://bugs.webkit.org/show_bug.cgi?id=218758</a>

Reviewed by Tim Nguyen.

This change ensures that interacting with (clicking/pressing/etc.) any
interactive descendants of a “summary” element will trigger the
activation behavior of those descendants themselves — rather than
instead triggering the “summary” element’s “details” parent to open.

Otherwise, without this change, the activation behavior of interactive
content within a “summary” element will get triggered only if that
interactive content is a direct child of the “summary” element.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-summary-element/anchor-with-inline-element-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-summary-element/interactive-content-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-summary-element/interactive-content.html: Added.
* Source/WebCore/html/HTMLSummaryElement.cpp:
(WebCore::isInSummaryInteractiveContent):
(WebCore::HTMLSummaryElement::defaultEventHandler):
(WebCore::isClickableControl): Deleted.

Canonical link: <a href="https://commits.webkit.org/267491@main">https://commits.webkit.org/267491@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c4e94017e6fced56d17e4e417651a06ceb904de3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/16739 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/17063 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/17513 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/18521 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/15681 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/16929 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/20295 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/17202 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/17972 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/16938 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/17308 "2 new passes 2 flakes 1 failures") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/14482 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/19310 "Built successfully") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14552 "4 flakes 2 failures") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/15166 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/21920 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15541 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/15332 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/19633 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/15941 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/13520 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/15118 "Built successfully") | [❌ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/15006 "Exiting early after 10 failures. 12 tests run. 1 flakes") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4011 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19483 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/15771 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/html/HTMLSummaryElement.cpp:
```diff
@@ -98,10 +100,30 @@ bool HTMLSummaryElement::isActiveSummary() const
 
 static bool isClickableControl(EventTarget* target)
```
It would be nice to get a link to the spec defining this function

sideshowbarker replied:
>> ```c++
>> static bool isClickableControl(EventTarget* target)
>> ```
> It would be nice to get a link to the spec defining this function

The spec doesn’t actually define “clickable control”. What it does define is “interactive content”, at https://html.spec.whatwg.org/multipage/dom.html#interactive-content.

Should I maybe rename the function to `isInteractiveContent(…)`?

nt1m replied:
`isInteractiveContentWithinSummary` would work.

sideshowbarker replied:
> `isInteractiveContentWithinSummary` would work.

I went with `isInInteractiveContentWithinSummary` — since the target might not itself be interactive content, but just needs to be a descendant of interactive content.

If you prefer making it simply `isInteractiveContentWithinSummary` though, I’m happy to change it to that.

nt1m commented on Source/WebCore/html/HTMLSummaryElement.cpp:
```diff
@@ -98,10 +100,30 @@ bool HTMLSummaryElement::isActiveSummary() const
+                return false;
+            if (ancestor.isInteractiveContent())
+                return true;
+        }
```
You can use `lineageOfType` instead of having duplicate checks for the current element and the ancestors.

nt1m commented on Source/WebCore/html/HTMLSummaryElement.cpp:
```diff
@@ -101,7 +103,23 @@ static bool isClickableControl(EventTarget* target)
+                return true;
+        }
+        return false;
+    }
```
What should happen if you have SVG in HTML or vice-versa? Would be good to have a spec reference to clarify those things :)

sideshowbarker replied:
> What should happen if you have SVG in HTML or vice-versa?

The only SVG element that has activation behavior is the SVG `a` element. The HTML spec doesn’t explicitly state that, because the HTML spec doesn’t state anything explicitly about the behavior of particular SVG elements.

But the only way you can have an SVG `a` element inside a `summary` is if there’s a `svg` element descendant of the `summary`, so that sets the SVG namespace, and then an SVG `a` descendant of that.

As far as HTML in SVG, I think the only relevant way that could be done in this context is by having an `svg` descendant of the `summary` and then a `foreignObject` descendant of that `svg` element, and then some HTML interactive content (with activation behavior) inside that `foreignObject`.

That `foreignObject` case seemed to me like too much of a pathological case to try to deal with — it’s hard to imagine any legitimate real-world need to ever mark up some `summary` content using that.

But if we want to handle the `foreignObject` case, I guess we could do that `lineageOfType` check using `Element` as the type, and then inside there, do an `is<HTMLElement>` type check of each ancestorOrSelf, and then a cast before calling `isInteractiveContent()`. It’d just be uglier and catching the `foreignObject` corner case would be the only reason for the additional complexity and runtime cost. 

But aside from that, what would help to simplify the existing code in the patch is if the spec explicitly defined the SVG `a` element as “interactive content”. I already chatted with Anne about that and I’ll be raising a spec PR tomorrow to make that spec change. (That change actually would also have normative implementation impact on `label` handling.) But the current patch in this branch would get us passing all the relevant WPT tests — without regressing anything — so I was thinking (hoping) we could possibly land this first, and deal with implementing the SVG-`a`-element-as-interactive-content change after this has landed first, and after the spec change lands.

> Would be good to have a spec reference to clarify those things :)

The spec isn’t explicit about the requirements for how interactive content works when inside `summary`. It seems to only be defined implicitly, by the places in the spec that state what the “activation behavior” is for any elements that have “activation behavior”. And what’s implicit about it is: the activation behavior of elements that are descendants of `summary` take precedence over the activation behavior of `summary` itself. I guess ideally the spec should be updated to state that explicitly.

nt1m replied:
OK, I'm fine addressing this in a follow up then, since this is interoperable. Regarding the spec, I was hoping the summary activation behavior would reflect this check too: https://html.spec.whatwg.org/multipage/interactive-elements.html#the-summary-element

sideshowbarker replied:
> Regarding the spec, I was hoping the summary activation behavior would reflect this check too: [html.spec.whatwg.org/multipage/interactive-elements.html#the-summary-element](https://html.spec.whatwg.org/multipage/interactive-elements.html#the-summary-element)

I’m happy to add that, but shall we do handle it with a separate follow-up PR? For one thing, it seems we may not have any tests for it — so I reckon I’d also need to write tests for it.

nt1m replied:
Yeah that's fine!

sideshowbarker replied:
> What should happen if you have SVG in HTML or vice-versa? Would be good to have a spec reference to clarify those things :)

The tests I added now include testing the HTML-in-SVG case — inside `foreignObject`. And in the existing tests we already had tests for the SVG-in-HTML case. So it seems we got coverage for all the SVG-HTML combination cases now (and the patch in this PR branch is handling them all as expected).

nt1m commented on Source/WebCore/html/HTMLSummaryElement.cpp:
```diff
@@ -101,7 +103,23 @@ static bool isClickableControl(EventTarget* target)
     if (!is<Element>(target))
         return false;
     auto& element = downcast<Element>(*target);
-    return is<HTMLFormControlElement>(element) || is<HTMLFormControlElement>(element.shadowHost());
```
The old code checks the shadowHost too (and Chromium's code also seems to iterate over shadowHosts as well fwiw). Not sure this matters.

sideshowbarker replied:
> The old code checks the shadowHost too (and Chromium's code also seems to iterate over shadowHosts as well fwiw). Not sure this matters.

Yeah, I wondered about that. It’s not clear to me what the spec says about checking the shadowHost. It seems like another place where ideally we’d want to have some clarification added to the spec. If we get that clarification made to the spec — or if the spec is already clear on this, but I haven’t been looking in the right place — I’m happy to do a separate follow-up PR. But again this seems like another case where we may not have tests for it yet (in which case I’d need to write them, along with the WebKit patch).

nt1m replied:
Sounds good, just wanted to make sure we're aware of the extent of the changes here :)

sideshowbarker replied:
> The old code checks the shadowHost too (and Chromium's code also seems to iterate over shadowHosts as well fwiw). Not sure this matters.

So it turns out that it does matter — for the cases of `textarea` and `input type=text` and I assume for other `input` types as well. (`textarea` and `input type=text` are the ones we have tests for.)

In those cases, when you click with them, the events get fired not at those elements themselves, but instead at `div` elements they host in the UA shadow DOM.

nt1m commented on Source/WebCore/html/HTMLSummaryElement.cpp:
```diff
@@ -96,12 +98,28 @@ bool HTMLSummaryElement::isActiveSummary() const
 }
 
-static bool isClickableControl(EventTarget* target)
+static bool isInInteractiveContentWithinSummary(EventTarget* target)
```
`isInSummaryInteractiveContent` ?

sideshowbarker replied:
Thanks — made it `isInSummaryInteractiveContent`

nt1m commented on Source/WebCore/html/HTMLSummaryElement.cpp:
```diff
@@ -96,12 +98,28 @@ bool HTMLSummaryElement::isActiveSummary() const
+                return true;
+        }
+        return false;
+    }
```
Can we write this in a single loop?

```suggestion
    for (auto* element = &startElement; element; element = element->parentOrShadowHostNode()) {
        if (is<HTMLSummaryElement>(element))
            return false;
        if (is<HTMLElement>(element) && downcast<HTMLElement>(element).isInteractiveContent())
            return true;
        if (is<SVGAElement>(element))
            return false;
    }
```

sideshowbarker replied:
> Can we write this in a single loop?

Thanks much — made it so (but with a Node rather than Element, since the Element `parentOrShadowHostNode` is private).

By the way, should I use `RefPtr<ContainerNode> node` here rather than `auto* node`?

nt1m reviewed (commented):
Can we add WPT coverage for the `<input>/<textarea>` case (potentially one with a shadow DOM like a date input)? You might need to use test_driver actions to emulate a click inside the date input.

nt1m commented on Source/WebCore/html/HTMLSummaryElement.cpp:
```diff
@@ -96,12 +98,15 @@ bool HTMLSummaryElement::isActiveSummary() const
         return false;
-    auto& element = downcast<Element>(*target);
-    return is<HTMLFormControlElement>(element) || is<HTMLFormControlElement>(element.shadowHost());
+    for (auto* node = &downcast<Node>(*target); node && !is<HTMLSummaryElement>(node); node = node->parentOrShadowHostNode()) {
```
I realize we have `parentOrShadowHostElement` which is probably more appropriate:

```suggestion
    for (auto* element = downcast<Element>(target); element && !is<HTMLSummaryElement>(element); element = element->parentOrShadowHostElement()) {
```

nt1m commented on Source/WebCore/html/HTMLSummaryElement.cpp:
```diff
@@ -96,12 +98,15 @@ bool HTMLSummaryElement::isActiveSummary() const
-    auto& element = downcast<Element>(*target);
-    return is<HTMLFormControlElement>(element) || is<HTMLFormControlElement>(element.shadowHost());
+    for (auto* element = downcast<Element>(target); element && !is<HTMLSummaryElement>(element); element = element->parentOrShadowHostElement()) {
+        if ((is<HTMLElement>(node) && downcast<HTMLElement>(node)->isInteractiveContent()) || is<SVGAElement>(node))
```
```suggestion
    for (RefPtr element = downcast<Element>(target); element && !is<HTMLSummaryElement>(element); element = element->parentOrShadowHostElement()) {
        if ((is<HTMLElement>(element) && downcast<HTMLElement>(element)->isInteractiveContent()) || is<SVGAElement>(element))
```

nt1m reviewed (approved):
Looks good, but please squash commits in order to enable landing.

Thanks for working on this!

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-summary-element/interactive-content.html:
```diff
@@ -0,0 +1,118 @@
+      bubbles: true,
+      cancelable: true
+    }));
+  }
```
We never assert the details state for the SVG elements.

More-over, I think this should use `await test_driver.click(target);` to emulate an actual native click on the shadow DOM of the input.

sideshowbarker replied:
OK — working on making those test update right now

sideshowbarker committed:
Don’t open “details” when interacting with interactive “summary” descendants
https://bugs.webkit.org/show_bug.cgi?id=218758

Reviewed by Tim Nguyen.

This change ensures that interacting with (clicking/pressing/etc.) any
interactive descendants of a “summary” element will trigger the
activation behavior of those descendants themselves — rather than
instead triggering the “summary” element’s “details” parent to open.

Otherwise, without this change, the activation behavior of interactive
content within a “summary” element will get triggered only if that
interactive content is a direct child of the “summary” element.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-summary-element/anchor-with-inline-element-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-summary-element/interactive-content-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-summary-element/interactive-content.html: Added.
* Source/WebCore/html/HTMLSummaryElement.cpp:
(WebCore::isInSummaryInteractiveContent):
(WebCore::HTMLSummaryElement::defaultEventHandler):
(WebCore::isClickableControl): Deleted.

Canonical link: https://commits.webkit.org/267491@main
